flink-table/flink-table-api-java/src/main/java/org/apache/flink/table/delegation/Parser.java

```diff
@@ -0,0 +1,57 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.flink.table.delegation;
+
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.table.catalog.UnresolvedIdentifier;
+import org.apache.flink.table.operations.Operation;
+import org.apache.flink.table.operations.QueryOperation;
+
+import java.util.List;
+
+/**
+ * Provides methods for parsing SQL objects from a SQL string.
+ */
+@Internal
+public interface Parser {
+
+	/**
+	 * Entry point for parsing SQL queries expressed as a String.
+	 *
+	 * <p><b>Note:</b>If the created {@link Operation} is a {@link QueryOperation}
+	 * it must be in a form that will be understood by the
+	 * {@link Planner#translate(List)} method.
+	 *
+	 * <p>The produced Operation trees should already be validated.
+	 *
+	 * @param statement the SQL statement to evaluate
+	 * @return parsed queries as trees of relational {@link Operation}s
+	 * @throws org.apache.flink.table.api.SqlParserException when failed to parse the statement
+	 */
+	List<Operation> parse(String statement);
+
+	/**
+	 * Entry point for parsing SQL identifiers expressed as a String.
+	 *
+	 * @param identifier the SQL identifier to parse
+	 * @return parsed identifier
+	 * @throws org.apache.flink.table.api.SqlParserException when failed to parse the identifier
+	 */
+	UnresolvedIdentifier parseIdentifier(String identifier);
```

Comment:
IIUC when user enters a table name / function name, we will translate it through following path:
`string ----> UnresolvedIdentifier(via Parser) ----> ObjectIdentifier(via CatalogManager)`

Besides with my previous comment in `UnresolvedIdentifier`, I'm not sure whether this is necessary. It also sounds not intuitive that it's parser's responsibility to convert the string to `UnresolvedIdentifier`. 

One alternative solution might be: 
1. we can relax the ObjectIdentifier to tolerant partial identifies, i.e. only table name. Then we can pass `String` or `String[]` to ObjectIdentifier to construct. 
2. ObjectIdentifier can tell whether itself has been qualified or not. And we can qualify the ObjectIdentifier through `CatalogManager`.

Reply:
That's exactly what I don't want to do. I really aim for having a clear separation between resolved/qualified objects and non-qualified ones.

I really don't want to relax the restrictions on ObjectIdentifier as it makes it much harder to use. Now we can just rely on the fact that the ObjectIdentifier always has three parts sets whenever we use it. During objects lookup/serialization etc.

Reply:
I'm ok with `UnresolvedIdentifier` now. How about the interface about `parseIdentifier`, what's your thought about moving it to `CatalogManager`?

Reply:
Why do you think it's not intuitive for a parser to create an UnresolvedIdentifier? 

The alternative would be to return a String[], but I think it's cleaner to return a proper object rather than string's array. This way it already aligns the array with catalog/database/object.

Reply:
After some more thoughts, I think this API is also reasonable. My first impression about this is, normally Parser have to deal with a whole SQL string, and table/function identifies are among the SQL string. But I missed other scenarios like when people register or delete table/function, the user will provide the identify directly.

Reply:
Cool!

Yes I agree that usually we should parse the whole string, there will be cases where we need to parse just subexpressions though e.g. computed columns. The identifier is the first one.